### PR TITLE
Update handling of dates to be more robust

### PR DESCRIPTION
**Pull request by seanmanson**
The current validators for MaxDate/MinDate fail to handle invalid dates correctly, throwing exceptions when they shouldn't when `$value` is invalid. I updated the validators so this doesn't happen, added tests, and added a format parameter so max/min dates can be specified for a given date format (`max_date:07/08/2009,d/m/Y`)

I had to update phpunit to get it to work with my version of php...

### CODE REVIEW

File: src/IntelliHR/Validation/Validators/AbstractValidator.php
Comment (EspadaV8) [inline src/IntelliHR/Validation/Validators/AbstractValidator.php]: This can be type hinted

File: src/IntelliHR/Validation/Validators/AbstractValidator.php
Comment (EspadaV8) [inline src/IntelliHR/Validation/Validators/AbstractValidator.php]: When does this return a `bool`?

File: src/IntelliHR/Validation/Validators/MaxDate.php
Comment (EspadaV8) [inline src/IntelliHR/Validation/Validators/MaxDate.php]: Since you're making a lot of changes around here, I'd replace the usage of `DateTime` objects with `Carbon` objects instead and then use `$date->lte($maxDate)`. What is actually happening here is the `<=` is converting the 2 dates to a string and then comparing them as string values.

File: src/IntelliHR/Validation/Validators/MinDate.php
Comment (EspadaV8) [inline src/IntelliHR/Validation/Validators/MinDate.php]: Similar to the other class, I'd change this to make use of `Carbon` and `gte`

**Review by EspadaV8 — CHANGES_REQUESTED** [review]

File: src/IntelliHR/Validation/Validators/AbstractValidator.php
Comment (seanmanson) [reply src/IntelliHR/Validation/Validators/AbstractValidator.php]: PHP returns `false` for some invalid dates, and throws an error for others 😕

File: src/IntelliHR/Validation/Validators/MaxDate.php
Comment (btaylor-ihr) [inline src/IntelliHR/Validation/Validators/MaxDate.php:29]: Would this also need to be in the try/catch?

File: src/IntelliHR/Validation/Validators/MaxDate.php
Comment (btaylor-ihr) [inline src/IntelliHR/Validation/Validators/MaxDate.php]: Not sure if I'm reading it correctly, but If the date fails to be created with the specified format, you're passing validation?

Comment (btaylor-ihr) [reply src/IntelliHR/Validation/Validators/MaxDate.php]: Oh so it just returns false if it fails? In which case it still passes validation anyways, so my question is still relevant

Comment (btaylor-ihr) [reply src/IntelliHR/Validation/Validators/MaxDate.php]: Maybe get `getDateForParameter` to throw the exception for you if it's going to return `false`. Makes it a bit easier to read. Because yay, PHP nuances

File: tests/Validators/MaxDateTest.php
Comment (btaylor-ihr) [inline tests/Validators/MaxDateTest.php]: Is importing a class considered a styling thing? If so, disregard

File: tests/Validators/MaxDateTest.php
Comment (btaylor-ihr) [inline tests/Validators/MaxDateTest.php:95]: Ok, so now it definitely looks like you're wanting bad dates to pass. Why is that?

File: src/IntelliHR/Validation/Validators/MaxDate.php
Comment (seanmanson) [reply src/IntelliHR/Validation/Validators/MaxDate.php]: Yes, it should pass validation if the value is invalid. I think it's the job of a separate `date_format` validator to test if the date is invalid, not `max_date` (otherwise you'd end up with two validation errors when only one of them is applicable)

File: src/IntelliHR/Validation/Validators/MaxDate.php
Comment (seanmanson) [reply src/IntelliHR/Validation/Validators/MaxDate.php:29]: I want invalid parameters to still throw an exception, as that's a code issue rather than a validation issue

File: tests/Validators/MaxDateTest.php
Comment (seanmanson) [reply tests/Validators/MaxDateTest.php:95]: >It should pass validation if the value is invalid. I think it's the job of a separate `date_format` validator to test if the date is invalid, not `max_date` (otherwise you'd end up with two validation errors when only one of them is applicable)

File: src/IntelliHR/Validation/Validators/MaxDate.php
Comment (btaylor-ihr) [reply src/IntelliHR/Validation/Validators/MaxDate.php:29]: Fair enough. This behaviour is tested from memory

File: tests/Validators/MaxDateTest.php
Comment (btaylor-ihr) [reply tests/Validators/MaxDateTest.php:95]: >Fair enough. This behaviour is tested from memory

TIL quoting. Thanks :p

**Review by btaylor-ihr — APPROVED** [review]
I'm adding my approval to this, but it's up to @EspadaV8 to merge as he'd requested changes

File: composer.json
Comment (EspadaV8) [inline composer.json]: Now that it's using `Carbon` shouldn't it be required here?

Comment (EspadaV8) [reply composer.json]: I've just noticed that the `composer.lock` file includes Laravel 5.5.2, this will prevent the package from being installed. I think the version constraint here is incorrect and we should probably have `^5.1` (IIRC).

**Review by EspadaV8 — CHANGES_REQUESTED** [review]
The code seems fine. Just need to check that version of Laravel that's required and possibly add Carbon as a requirement too.

File: composer.json
Comment (seanmanson) [reply composer.json]: It was already being included as a dependency of this, but I'll add it separately too

**Review by EspadaV8 — COMMENTED** [review]
Just need to update the `composer.json` file to make sure that only Laravel 5.1 components are used and then update the `composer.lock` file too. After that it looks okay to merge in

**Pull request merged**
